haskell/src/Main.hs

```diff
@@ -0,0 +1,170 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE MultiWayIf #-}
+module Main where
+
+import UI.NCurses
+import Control.Monad.IO.Class(liftIO)
+import Control.Monad(unless)
+import System.Environment(getArgs)
+import Data.List(sort)
+import Prelude hiding (Either(..))
+
+type Coord = (Int, Int)
+
+data Move = Up | Left | Right | Down
+
+data Level = Level { walls :: [Coord]
+                   , storage :: [Coord]
+                   , crates :: [Coord]
```

Comment:
After looking at the functions checking type below, I wonder if there's another data type you could use to capture these.

```haskell
data Obstacle
    = Wall Coord
    | Storage Coord StorageContents -- obv StorageContents doesn't exist but that'd be pretty neat
    | Crate Coord SlideDistance -- maybe some crates are slippery and move two spaces?
```

With this, `Level` would have a list of obstacles, each with a `Coord`, `performMove` could check if there's an obstacle in the space they're attempting to move to and pattern-match on the type to dictate movement outcome.